broker.go

```diff
@@ -0,0 +1,645 @@
+// from the same peer are handled serially.  If a handler is busy handling a
+// message, and more messages arrive from the same peer, then the last message
+// replaces the previous unhandled message to avoid having to maintain queues
+// of messages.  Handlers do not have persistent goroutines, but start a new
+// goroutine to handle a single message.
+type Broker struct {
+	gsExchange graphsync.GraphExchange
+	// dss captures the default selector sequence passed to
+	// ExploreRecursiveWithStopNode.
+	dss         ipld.Node
+	dtManager   dt.Manager
+	tmpDir      string
+	unsubEvents dt.Unsubscribe
+
+	addrTTL time.Duration
+	host    host.Host
+	psub    *pubsub.Subscription
+	topic   *pubsub.Topic
+
+	allowPeer     AllowPeerFunc
+	handlers      map[peer.ID]*handler
+	handlersMutex sync.Mutex
+
+	// Map of CID of in-progress sync to sync done channel.
+	syncDoneChans map[cid.Cid]chan<- struct{}
+	syncDoneMutex sync.Mutex
+
+	// inEvents is used to send a SyncFinished from a peer handler to the
+	// distributeEvents goroutine.
+	inEvents chan SyncFinished
+
+	// outEventsChans is a slice of channels, where each channel delivers a
+	// copy of a SyncFinished to an OnSyncFinished reader.
+	outEventsChans []chan SyncFinished
+	outEventsMutex sync.Mutex
+
+	// closing signals that the Broker is closing.
+	closing chan struct{}
+	// cancel sub-services (pubsub, datatransfer).
+	cancelAll context.CancelFunc
+	// closeOnde ensures that the Close only happens once.
+	closeOnce sync.Once
+}
+
+// SyncFinished notifies an OnSyncFinished reader that a specified peer
+// completed a sync.  The channel receives events from providers that are
+// manually synced to the latest, as well as those auto-discovered.
+type SyncFinished struct {
+	// Cid is the CID identifying the link that finished and is now the latest
+	// sync for a specific peer.
+	Cid cid.Cid
+	// PeerID identifies the peer this SyncFinished event pertains to.
+	PeerID peer.ID
+}
+
+// handler holds state that is specific to a peer
+type handler struct {
+	dtManager dt.Manager
+	// distEvents is used to communicate a syncDone event back to the Broker
+	// for distribution to OnSyncFinished readers.
+	distEvents chan<- SyncFinished
+	latestSync ipld.Link
+	msgChan    chan cid.Cid
+	// peerID is the ID of the peer this handler is responsible for.
+	peerID peer.ID
+	// putSyncDone is a function supplied by the Broker for the handler to give
+	// Broker the channel to send sync done notification to.
+	putSyncDone func(cid.Cid, chan<- struct{})
+	// syncMutex serializes the handling of individual syncs
+	syncMutex sync.Mutex
+}
+
+// NewBroker creates a new Broker that process pubsub messages.
+func NewBroker(host host.Host, ds datastore.Batching, lsys ipld.LinkSystem, topic string, dss ipld.Node, options ...Option) (*Broker, error) {
+	cfg := config{
+		addrTTL: defaultAddrTTL,
+	}
+	err := cfg.apply(options)
+	if err != nil {
+		panic(err.Error())
```

Comment:
Why panic?

Reply:
Was thinking to ensure this was caught as a programming error.  Probably better just to return the error.